### PR TITLE
[feature] Enable on-device grammar sampling

### PR DESCRIPTION
**Pull request by sanjaradylov**
## PR Category

Feature

## TL;DR

Adds capability-gated grammar masking to TT on-device decode sampling. This enables structured outputs to remain on device when paired with the vLLM TT plugin.

## Usage

Quick overview:

```python
# --- capability gating (models/common/model_capabilities.py) ---
# instead of an allowlist by model name, generators declare support:
model_capabilities = {"supports_device_grammar": True, ...}

# --- generator.py: deferred, grammar-aware decode sampling ---
tt_logits = generator.decode_forward(...)          # forward only; no sampling yet
tokens = generator.sample_decode_on_device(
    tt_logits,
    sampling_params,
    grammar_bitmask=grammar_bitmask,   # NEW: packed torch.int32 mask, or None
    enable_trace=True,
)

# --- tt_sampling.py: packed grammar mask helpers (new) ---
from models.common.sampling.tt_sampling import format_grammar_bitmask

device_mask = format_grammar_bitmask(
    grammar_bitmask,             # vLLM's raw packed mask: [batch, ceil(vocab/32)]
    vocab_size=vocab_size,
    padded_vocab_size=padded_vocab_size,
    max_batch_size=max_batch_size,
)  # -> validated/padded [max_batch_size, ceil(padded_vocab/32)] int32 tensor

sampler.update_grammar_bitmask(device_mask)   # replaces the full mask (no incremental update)
```

vLLM serving:

```bash
python examples/server_example_tt.py \
  --model Qwen/Qwen2.5-7B-Instruct \
  --max_num_seqs 32 \
  --block_size 64 \
  --guided-decoding-backend xgrammar \
  --max_model_len 4096 \
  --additional-config '{"tt": {"sample_on_device_mode": "decode_only"}}'

curl http://localhost:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "Qwen/Qwen2.5-7B-Instruct",
    "messages": [{"role": "user", "content": "Return a JSON object with fields name and age."}],
    "temperature": 0.7,
    "top_p": 0.9,
    "top_k": 20,
    "max_tokens": 64,
    "guided_json": {
      "type": "object",
      "properties": {"name": {"type": "string"}, "age": {"type": "integer"}},
      "required": ["name", "age"]
    }
  }'
```

## Details

The generator can defer token selection until a packed grammar mask arrives, apply it on device, and then sample only allowed tokens. This enables structured outputs with lower host overhead and latency; initial support is decode-only and excludes async scheduling and logprobs.

- **`.github/workflows/vllm-model-tests-impl.yaml`** - Adds two-wave device-grammar path validation.
- **`models/common/model_capabilities.py`** - Adds the device-grammar capability.
- **`models/common/sampling/generator.py`** - Adds grammar-aware sampling traces and mask updates.
- **`models/common/sampling/tt_sampling.py`** - Validates, stores, unpacks, and applies packed grammar masks before sampling.
- **`models/common/tests/test_sampling.py`** - Adds host unit coverage for masks, traces, warmup, deferred sampling, failures, and slot remapping.
- **`models/common/tests/test_tt_sampling.py`** - Adds synthetic eager and traced device-mask tests, including request turnover.
- **`models/common/warmup/warmup_utils.py`** - Warms supported grammar sampling variants before trace capture.
- **`models/tt_transformers/tt/generator.py`** - Adds deferred decode sampling, runtime activation, failure poisoning, trace preparation, DP mask splitting, and sampler-state remapping.
- **`models/tt_transformers/tt/generator_vllm.py`** - Enables the capability for Llama, Qwen, and Mistral generators.
- **`tests/pipeline_reorg/vllm_model_tests.yaml`** - Configures the P150x4 Llama device-grammar test leg.

## Test Strategy

- Host unit tests cover mask validation and padding, trace isolation, warmup variants, deferred payload ownership, terminal failures, and slot remapping.
- Synthetic device unit tests cover eager masking, traced masking, and stale-mask removal after request turnover.
- Device test matrix: five cases collected across single- and multi-device configurations.

## Test Results

- [x] Host unit tests - **`models/common/tests/test_tt_sampling.py`**
  - Grammar bitmask packing/validation logic (bit semantics,  -1  all-allowed words, tail-bit clearing, empty-mask rejection): pass
  - New `grammar_shard_coverage_tokens`  helper, pinned for `(4,8)` : derives  4095, 4096, 18432, 31999  as shard-boundary/final-token representatives: pass
- [x] Device tests (one-chip, via MCP) - **`models/common/tests/test_tt_sampling.py::TestGrammarSampling`** 
  - Eager grammar sampling with shard-boundary tokens: pass
  - Traced grammar sampling with turnover (constrained → all-allowed → constrained), using corrected oracle rows: pass
  - `(4,8)` / 32-device-only cases: skipped (only 1 device available in this environment)
  - Result: 3 passed, 2 skipped 
- [ ] CI verification
  - Confirmed **`.github/workflows/vllm-model-tests-impl.yaml` and **`tests/pipeline_reorg/vllm_model_tests.yaml`** are byte-identical to `origin/main` after temporarily removing the paired-plugin-dependent step.
  - [Tier 1 Models Unit (all)](https://github.com/tenstorrent/tt-metal/actions/runs/34605644942)

Gaps:
- [ ] Full non-grammar **`test_tt_sampling.py`** suite (top-k/top-p/argmax/logprobs/force-argmax) - not run against this branch, so no regression evidence there yet
- [ ] `(4,8)` / 32-device grammar cases - not executed on real multi-device hardware (only skip-verified on 1 device)
- [ ] P150x4 CI leg - not run (leg is currently disabled pending the paired plugin PR)
- [ ] Broader `tt_transformers` generator/decode integration tests - not run
- [ ] Performance tests

## Limitations

- Decode only - structured prefill remains host-sampled because intermediate prompt rows must not advance device sampling state.
- No async scheduling - the deferred logits payload is single-owner; another forward cannot safely start before the current grammar mask arrives.
- No structured logprobs - grammar-aware device logprob output is not implemented, so these batches retain host sampling.
- No row-sharded sampling - standard data parallelism is supported, but `sampling_dp != data_parallel`  has a different row layout without a proven grammar-to-slot mapping.
- Llama, Qwen, and Mistral only - these generators share the sampler, deferred-decode, and trace contract implemented here; this is not a Metal hardware restriction.
- No block-output models - `DiffusionGemma` and similar models own a separate multi-token sampler that does not support this grammar contract.
- Batch-wide eligibility - constrained and unconstrained requests can share one device-sampled batch; however, one request requiring an unsupported host-only feature moves the whole batch to host sampling
- Failures are terminal - after deferred decode starts, device state may have advanced, so an error cannot safely retry on host and requires runner reconstruction.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:sanjaradylov/device-grammar-sampling)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:sanjaradylov/device-grammar-sampling)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:sanjaradylov/device-grammar-sampling)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:sanjaradylov/device-grammar-sampling)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:sanjaradylov/device-grammar-sampling)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:sanjaradylov/device-grammar-sampling)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:sanjaradylov/device-grammar-sampling)

### CODE REVIEW

File: models/common/warmup/warmup_utils.py [hunk 1]
```diff
@@ -128,8 +163,38 @@ def warmup_model_decode(
...
+                # The common generator uses the first traced call to allocate
+                # the persistent decode feedback buffer and precompile every
+                # sampler variant against that exact buffer before recording
+                # any trace. End with this call's params so its trace key
+                # matches the state captured immediately afterwards.
+                decode_kwargs["sampling_trace_warmup_params"] = (
+                    sampling_params[index + 1 :] + sampling_params[: index + 1]
+                )
+                decode_kwargs["sampling_trace_warmup_device_grammar"] = can_sample_device_grammar
+                trace_variants_prepared = True
```
[the remark this answers is not in the document]
Comment (sanjaradylov) [reply models/common/warmup/warmup_utils.py:176]: This duplicate precompile does not occur. `_prepare_decode_trace_text()` precompiles the sampler variants before recording the model trace, and `_record_decode_trace_text()` unconditionally calls `sampling_module.capture_trace(..., skip_precompile=True)`. The absence of `skip_trace_precompile` on the first outer `decode_forward()` call therefore does not cause ёSamplingGenerator.capture_trace()ё to precompile again. Subsequent traced grammar warmup calls also set the skip flag explicitly. The capture phase consequently does not perform another allocation-producing sampler precompile.

**Comment by viktorpusTT** [comment]
**1. Use local vocabulary widths for the sharded grammar-mask reshapes**

Primary location: [TTSampling._apply_grammar_bitmask, lines 549-552](https://github.com/tenstorrent/tt-metal/blob/904da9d35478e70137f4d6a81fadbf140140ecb7/models/common/sampling/tt_sampling.py#L549-L552). The second reshape at [lines 566-569](https://github.com/tenstorrent/tt-metal/blob/904da9d35478e70137f4d6a81fadbf140140ecb7/models/common/sampling/tt_sampling.py#L566-L569) needs the same correction.

`TTSampling` applies structured-output constraints by expanding each packed `int32` grammar word into 32 allowed-token values. Multi-device sampling partitions the vocabulary across a mesh axis, so each device owns only part of the packed mask. `TTSampling._apply_grammar_bitmask` nevertheless uses the global `self.padded_vocab_size` for both reshapes. TTNN requires each reshape to preserve the local tensor volume.

The failure occurs whenever grammar sampling uses more than one vocabulary shard. With the paired plugin, an eligible Llama, Qwen, or Mistral configuration can fail during grammar warmup before the server accepts requests.

Correction: derive the packed width from `self.grammar_bitmask_tensor.shape[-1]`. Derive the unpacked width from that local width multiplied by 32, and verify the unpacked width matches the local logits width. Correct both reshapes. Run grammar tests on a mesh with multiple vocabulary shards.

**2. Preserve the Gemma4 sampler wrapper signature**

Primary location: [Generator.sample_decode_on_device, line 2719](https://github.com/tenstorrent/tt-metal/blob/904da9d35478e70137f4d6a81fadbf140140ecb7/models/tt_transformers/tt/generator.py#L2714-L2720).

Gemma4 uses `ChunkedPrefillPageTableGuardMixin.sample_decode_on_device` to insert a persistent output-token buffer into eager sampling. The Gemma4 wrapper temporarily replaces `sampling_module.sample`, then calls the shared `Generator.sample_decode_on_device`. PR #55892 makes the shared implementation pass `grammar_bitmask` on every call, including ordinary sampling with `grammar_bitmask=None`. The Gemma4 `_sample` wrapper does not accept `grammar_bitmask`, so Python raises `TypeError` before sampling.

Both [Gemma4ForCausalLM](https://github.com/tenstorrent/tt-metal/blob/904da9d35478e70137f4d6a81fadbf140140ecb7/models/demos/gemma4/tt/generator_vllm.py#L209) and [Gemma4Generator](https://github.com/tenstorrent/tt-metal/blob/904da9d35478e70137f4d6a81fadbf140140ecb7/models/demos/gemma4/tt/generator.py#L1577) inherit `ChunkedPrefillPageTableGuardMixin`. Non-PLI Gemma4 configurations with a constructed device sampler and normal device token feedback are affected in serving and demo execution. Structured-output requests are not required. 

Correction: extend the Gemma4 `_sample` wrapper to accept and forward `grammar_bitmask`, or add `grammar_bitmask` to the shared call only when a grammar mask is present. Add a host regression through the real Gemma4 mixin with feedback enabled. The new `_StubSamplingModule.sample` accepts `**kwargs` at [test_sampling.py:693](https://github.com/tenstorrent/tt-metal/blob/904da9d35478e70137f4d6a81fadbf140140ecb7/models/common/tests/test_sampling.py#L693), so the stub cannot detect the wrapper incompatibility.
